File: app/src/main/java/com/rakuishi/coinchecker/Currency.java
package com.rakuishi.coinchecker;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;

public class Currency {
    int id;
    String name;
    String unit;

    private Currency(int id, String name, String unit) {
        this.id = id;
        this.name = name;
        this.unit = unit;
    }

    public String getPair() {
        return unit + "_jpy";
    }

    @DrawableRes
    public int getIconResId(Context context) {
        return context.getResources().getIdentifier("icon_" + unit, "drawable", context.getPackageName());
    }

    // region pref

    private static final String PREFS_NAME = "com.rakuishi.coinchecker.currency";
    private static final String PREF_PREFIX_KEY = "pref_";

    public static void saveCurrencyPref(Context context, int appWidgetId, int currencyId) {
        SharedPreferences.Editor prefs = context.getSharedPreferences(PREFS_NAME, 0).edit();
        prefs.putInt(PREF_PREFIX_KEY + appWidgetId, currencyId);
        prefs.apply();
    }

    public static Currency loadCurrencyPref(Context context, int appWidgetId) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, 0);
        int currencyId = prefs.getInt(PREF_PREFIX_KEY + appWidgetId, 0);
        return getCurrency(currencyId);
    }

    public static void deleteCurrencyPref(Context context, int appWidgetId) {
        SharedPreferences.Editor prefs = context.getSharedPreferences(PREFS_NAME, 0).edit();
        prefs.remove(PREF_PREFIX_KEY + appWidgetId);
        prefs.apply();
    }

    // endregion

    // region defined currencies

    public static ArrayList<Currency> getCurrencies() {
        ArrayList<Currency> coins = new ArrayList<>();
        for (int id = 0; id < 12; id++) {
            coins.add(getCurrency(id));
        }
        return coins;
    }

    @Nullable
    public static Currency getCurrency(int id) {
        switch (id) {
            case 0:
                return new Currency(0, "Bitcoin", "btc");
            case 1:
                return new Currency(1, "Ethereum", "eth");
            case 2:
                return new Currency(2, "Ethereum Classic", "etc");
            case 3:
                return new Currency(3, "LISK", "lsk");
            case 4:
                return new Currency(4, "Factom", "fct");
            case 5:
                return new Currency(5, "Monero", "xmr");
            case 6:
                return new Currency(6, "Augur", "rep");
            case 7:
                return new Currency(7, "Ripple", "xrp");
            case 8:
                return new Currency(8, "Zcash", "zec");
            case 9:
                return new Currency(9, "NEM", "xem");
            case 10:
                return new Currency(10, "Litecoin", "ltc");
            case 11:
                return new Currency(11, "DASH", "dash");
            default:
                return null;
        }
    }

    // endregion
}

File: app/src/main/java/com/rakuishi/coinchecker/Currency.kt
package com.rakuishi.coinchecker

import android.content.Context
import android.support.annotation.DrawableRes
import java.util.*

class Currency private constructor(internal var id: Int, internal var name: String, internal var unit: String) {

    val pair: String
        get() = unit + "_jpy"

    @DrawableRes
    fun getIconResId(context: Context): Int {
        return context.resources.getIdentifier("icon_" + unit, "drawable", context.packageName)
    }

    companion object {

        // region pref

        private val PREFS_NAME = "com.rakuishi.coinchecker.currency"
        private val PREF_PREFIX_KEY = "pref_"

        fun saveCurrencyPref(context: Context, appWidgetId: Int, currencyId: Int) {
            val editor = context.getSharedPreferences(PREFS_NAME, 0).edit()
            editor.putInt(PREF_PREFIX_KEY + appWidgetId, currencyId)
            editor.apply()
        }

        fun loadCurrencyPref(context: Context, appWidgetId: Int): Currency? {
            val prefs = context.getSharedPreferences(PREFS_NAME, 0)
            val currencyId = prefs.getInt(PREF_PREFIX_KEY + appWidgetId, 0)
            return getCurrency(currencyId)
        }

        fun deleteCurrencyPref(context: Context, appWidgetId: Int) {
            val editor = context.getSharedPreferences(PREFS_NAME, 0).edit()
            editor.remove(PREF_PREFIX_KEY + appWidgetId)
            editor.apply()
        }

        // endregion

        // region defined currencies

        val currencies: ArrayList<Currency>
            get() {
                val coins = ArrayList<Currency>()
                for (id in 0..11) {
                    getCurrency(id)?.let { coins.add(it) }
                }
                return coins
            }

        fun getCurrency(id: Int): Currency? {
            when (id) {
                0 -> return Currency(0, "Bitcoin", "btc")
                1 -> return Currency(1, "Ethereum", "eth")
                2 -> return Currency(2, "Ethereum Classic", "etc")
                3 -> return Currency(3, "LISK", "lsk")
                4 -> return Currency(4, "Factom", "fct")
                5 -> return Currency(5, "Monero", "xmr")
                6 -> return Currency(6, "Augur", "rep")
                7 -> return Currency(7, "Ripple", "xrp")
                8 -> return Currency(8, "Zcash", "zec")
                9 -> return Currency(9, "NEM", "xem")
                10 -> return Currency(10, "Litecoin", "ltc")
                11 -> return Currency(11, "DASH", "dash")
                else -> return null
            }
        }
    }

    // endregion
}
